File: app/backend/src/routes/deployments.ts
import {Router, Response} from 'express';
import {AuthenticatedRequest} from '../types';
import {requireAuth, requireRole} from '../middleware/auth';
import * as k8s from '../services/kubernetes';
import {getArgoHealthMap} from '../services/argocd';
import {logger} from '../services/logger';
import {deploymentActions} from '../services/metrics';

const router = Router();
const NAME_RE = /^[a-z0-9]([a-z0-9\-.]*[a-z0-9])?$/;

async function validateTarget(
    res: Response,
    namespace: string,
    name: string,
): Promise<boolean> {
    if (!NAME_RE.test(namespace) || !NAME_RE.test(name)) {
        res.status(400).json({error: 'Invalid namespace or deployment name'});
        return false;
    }
    const managed = await k8s.discoverNamespaces();
    if (!managed.includes(namespace)) {
        res.status(403).json({error: 'Namespace not managed by this dashboard'});
        return false;
    }
    return true;
}

router.get('/', requireAuth, async (_req: AuthenticatedRequest, res: Response) => {
        try {
            const [deployments, argoMap] = await Promise.all([
                k8s.listDeployments(),
                getArgoHealthMap(),
            ]);

            for (const dep of deployments) {
                if (dep.argoApp && argoMap.has(dep.argoApp)) {
                    const a = argoMap.get(dep.argoApp)!;
                    dep.healthStatus = a.health;
                    dep.syncStatus = a.sync;
                }
            }

            res.json(deployments);
        } catch (err) {
            logger.error('Failed listing deployments', {error: err});
            res.status(500).json({error: 'Failed to list deployments'});
        }
    },
);

// Write operations require admin role
router.post('/:namespace/:name/restart', requireAuth, requireRole('admin'),
    async (req: AuthenticatedRequest, res: Response) => {
        const {namespace, name} = req.params;
        if (!(await validateTarget(res, namespace, name))) return;

        try {
            await k8s.restartDeployment(namespace, name);
            deploymentActions.inc({action: 'restart', namespace, deployment: name});
            logger.info('Restart requested', {
                namespace,
                name,
                user: req.user?.email,
            });
            res.json({ok: true});
        } catch (err) {
            logger.error('Restart failed', {namespace, name, error: err});
            res.status(500).json({error: 'Restart failed'});
        }
    },
);

router.post('/:namespace/:name/pause', requireAuth, requireRole('admin'),
    async (req: AuthenticatedRequest, res: Response) => {
        const {namespace, name} = req.params;
        if (!(await validateTarget(res, namespace, name))) return;

        try {
            await k8s.pauseDeployment(namespace, name);
            deploymentActions.inc({action: 'pause', namespace, deployment: name});
            logger.info('Pause requested', {
                namespace,
                name,
                user: req.user?.email,
            });
            res.json({ok: true});
        } catch (err) {
            logger.error('Pause failed', {namespace, name, error: err});
            res.status(500).json({error: (err as Error).message});
        }
    },
);

router.post('/:namespace/:name/resume', requireAuth, requireRole('admin'),
    async (req: AuthenticatedRequest, res: Response) => {
        const {namespace, name} = req.params;
        if (!(await validateTarget(res, namespace, name))) return;

        try {
            await k8s.resumeDeployment(namespace, name);
            deploymentActions.inc({action: 'resume', namespace, deployment: name});
            logger.info('Resume requested', {
                namespace,
                name,
                user: req.user?.email,
            });
            res.json({ok: true});
        } catch (err) {
            logger.error('Resume failed', {namespace, name, error: err});
            res.status(500).json({error: (err as Error).message});
        }
    },
);

export default router;

File: app/backend/internal/handlers/deployments.go
package handlers

import (
	"net/http"
	"regexp"

	chilib "github.com/go-chi/chi/v5"

	"github.com/mephalrith/noodles/backend/internal/errs"
	"github.com/mephalrith/noodles/backend/internal/middleware"
	"github.com/mephalrith/noodles/backend/internal/respond"
	"github.com/mephalrith/noodles/backend/internal/services"
)

var nameRE = regexp.MustCompile(`^[a-z0-9]([a-z0-9\-.]*[a-z0-9])?$`)

func validateTarget(w http.ResponseWriter, r *http.Request, k8s *services.K8sService, namespace, name string) bool {
	if !nameRE.MatchString(namespace) || !nameRE.MatchString(name) {
		respond.Error(w, errs.InvalidTarget)
		return false
	}
	managed, err := k8s.IsManagedNamespace(r.Context(), namespace)
	if err != nil || !managed {
		respond.Error(w, errs.NamespaceNotManaged)
		return false
	}
	return true
}

func HandleListDeployments(k8s *services.K8sService, argo *services.ArgoCDService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		deployments, err := k8s.ListDeployments(r.Context())
		if err != nil {
			services.Logger.Error("Failed listing deployments", "error", err)
			respond.Error(w, errs.Internal("Failed to list deployments"))
			return
		}

		argoMap := argo.GetHealthMap()
		for i := range deployments {
			dep := &deployments[i]
			if dep.ArgoApp != "" {
				if status, ok := argoMap[dep.ArgoApp]; ok {
					dep.HealthStatus = status.Health
					dep.SyncStatus = status.Sync
				}
			}
		}

		respond.OK(w, deployments)
	}
}

func HandleRestartDeployment(k8s *services.K8sService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		namespace := chilib.URLParam(r, "namespace")
		name := chilib.URLParam(r, "name")
		if !validateTarget(w, r, k8s, namespace, name) {
			return
		}

		if err := k8s.RestartDeployment(r.Context(), namespace, name); err != nil {
			services.Logger.Error("Restart failed", "namespace", namespace, "name", name, "error", err)
			respond.Error(w, errs.Internal("Restart failed"))
			return
		}

		user := middleware.UserFromContext(r.Context())
		services.DeploymentActions.With(map[string]string{"action": "restart", "namespace": namespace, "deployment": name}).Inc()
		services.Logger.Info("Restart requested", "namespace", namespace, "name", name, "user", user.Email)

		respond.OK(w, map[string]bool{"ok": true})
	}
}

func HandlePauseDeployment(k8s *services.K8sService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		namespace := chilib.URLParam(r, "namespace")
		name := chilib.URLParam(r, "name")
		if !validateTarget(w, r, k8s, namespace, name) {
			return
		}

		if err := k8s.PauseDeployment(r.Context(), namespace, name); err != nil {
			services.Logger.Error("Pause failed", "namespace", namespace, "name", name, "error", err)
			respond.Error(w, errs.Internal(err.Error()))
			return
		}

		user := middleware.UserFromContext(r.Context())
		services.DeploymentActions.With(map[string]string{"action": "pause", "namespace": namespace, "deployment": name}).Inc()
		services.Logger.Info("Pause requested", "namespace", namespace, "name", name, "user", user.Email)

		respond.OK(w, map[string]bool{"ok": true})
	}
}

func HandleResumeDeployment(k8s *services.K8sService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		namespace := chilib.URLParam(r, "namespace")
		name := chilib.URLParam(r, "name")
		if !validateTarget(w, r, k8s, namespace, name) {
			return
		}

		if err := k8s.ResumeDeployment(r.Context(), namespace, name); err != nil {
			services.Logger.Error("Resume failed", "namespace", namespace, "name", name, "error", err)
			respond.Error(w, errs.Internal(err.Error()))
			return
		}

		user := middleware.UserFromContext(r.Context())
		services.DeploymentActions.With(map[string]string{"action": "resume", "namespace": namespace, "deployment": name}).Inc()
		services.Logger.Info("Resume requested", "namespace", namespace, "name", name, "user", user.Email)

		respond.OK(w, map[string]bool{"ok": true})
	}
}
